sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1704 @@
+  }
+
+  def removeFromHostStore(value: SpillableHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    val hostStore = synchronized {
+      Option(storesInternal).map(_.hostStore)
+    }
+    hostStore.foreach(_.remove(value))
+  }
+
+  def removeFromDiskStore(value: DiskHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    val maybeStores = synchronized {
+      Option(storesInternal).map(_.diskStore)
+    }
+    maybeStores.foreach(_.remove(value))
+  }
+
+}
+
+/**
+ * A bounce buffer wrapper class that supports the concept of acquisition.
+ *
+ * The bounce buffer is acquired exclusively. So any calls to acquire while the
+ * buffer is in use will block at `acquire`. Calls to `release` notify the blocked
+ * threads, and they will check to see if they can acquire.
+ *
+ * `close` is the interface to unacquire the bounce buffer.
+ *
+ * `release` actually closes the underlying DeviceMemoryBuffer, and should be called
+ * once at the end of the lifetime of the executor.
+ *
+ * @param dmb - actual cudf DeviceMemoryBuffer that this class is protecting.
+ */
+private[spill] case class DeviceBounceBuffer(var dmb: DeviceMemoryBuffer) extends AutoCloseable {
+  private var acquired: Boolean = false
+  def acquire(): DeviceBounceBuffer = synchronized {
+    while (acquired) {
+      wait()
+    }
+    acquired = true
+    this
+  }
+
+  private def unaquire(): Unit = synchronized {
+    acquired = false
+    notifyAll()
+  }
+
+  override def close(): Unit = {
+    unaquire()
+  }
+
+  def release(): Unit = synchronized {
+    if (acquired) {
+      throw new IllegalStateException(
+        "closing device buffer pool, but some bounce buffers are in use.")
+    }
+    if (dmb != null) {
+      dmb.close()
+      dmb = null
+    }
+  }
+}
+
+/**
+ * A bounce buffer pool with buffers of size `bufferSize`
+ *
+ * This pool returns instances of `DeviceBounceBuffer`, that should
+ * be closed in order to be reused.
+ *
+ * Callers should synchronize before calling close on their `DeviceMemoryBuffer`s.
+ */
+trait DeviceBounceBufferPool extends AutoCloseable {
+  def bufferSize: Long
+  def nextBuffer(): DeviceBounceBuffer
+}
+
+/**
+ * ChunkedPacker is an Iterator-like class that uses a cudf::chunked_pack to copy a cuDF `Table`
+ * to a target buffer in chunks. It implements a next method that takes a DeviceMemoryBuffer
+ * as an argument to be used for the copy.
+ *
+ * Each chunk is sized at most `bounceBuffer.getLength`, and the caller should cudaMemcpy
+ * bytes from `bounceBuffer` to a target buffer after each call to `next()`.
+ *
+ * @note `ChunkedPacker` must be closed by the caller as it has GPU and host resources
+ *       associated with it.
+ *
+ * @param table cuDF Table to chunk_pack
+ * @param bounceBufferPool bounce buffer pool to use during the lifetime of this packer.
+ */
+class ChunkedPacker(table: Table,
```

Comment:
Nit: This class could be in a separate file.  This source file is already quite large.  Similarly the device/host/disk spill handle/store code could be placed in separate files to keep things from getting too big in one place.

Reply:
Yes, @zpuller has mentioned the same thing. OK if we break up the code in a follow on?

Reply:
Followup for this is fine with me.